gax-java/gax-httpjson/src/main/java/com/google/api/gax/httpjson/HttpJsonClientCalls.java

```diff
@@ -134,19 +137,29 @@ public void onMessage(T message) {
 
     @Override
     public void onClose(int statusCode, HttpJsonMetadata trailers) {
-      if (!isMessageReceived) {
-        if (trailers == null || trailers.getException() == null) {
-          future.setException(
-              new HttpJsonStatusRuntimeException(
-                  statusCode,
-                  "Exception during a client call closure",
-                  new NullPointerException(
-                      "Both response message and response exception were null")));
-        } else {
-          future.setException(trailers.getException());
-        }
-      } else {
+      Preconditions.checkNotNull(trailers);
+      // onClose() in invoked via HttpJsonClientCallImpl's close() method and
+      // the trailer exceptions are set via parameters in there.
+      //
+      // A successful response will have a null exception in the trailers.
+      // The trailer's exception CAN be a DEADLINE_EXCEEDED exception set
+      // from the deadlineScheduler and we must return it as an exception.
+      //
+      // Note: An exception has the highest priority. We must respect the timeout
+      // value set in the RetrySettings even if the message barely misses the
+      // cutoff time.
+      if (trailers.getException() != null) {
```

Comment:
Why do we need to change this logic? Is there a case that `isMessageReceived` is true and `DEADLINE_EXCEEDED` exception is set?

Reply:
There is a (very small) possibility that this can happen:
1. Enqueue a message/ response for call
2. `notifyListeners()` that a message has been sent which updates `isMessageReceived` = true
3. `timeout()` runs and enqueues the `DEADLINE_EXCEEDED` exception
4. Last loop of `deliver()` runs executes the the `onclose()` method

There is now a state where both a message and `DEADLINE_EXCEEDED` have been received.

Reply:
This is a flag to me that we may want to change our implementation, as now we have two source of truth to determine if we received response or not, `isMessageReceived` and `trailers.getException()`. Let's discuss more in our meeting.

Reply:
Per internal discussion, we've migrated this back to our old implementation.